rvar/src/Data/RVar.hs

```diff
@@ -1,112 +1,115 @@
-        , getRandomNByteInteger
-        )
-    
-    , RVar
-    , runRVar, sampleRVar
-    
+    ( RVar
+    , runRVar, sampleReaderRVar, sampleStateRVar
+    , pureRVar
+
     , RVarT
-    , runRVarT, sampleRVarT
-    , runRVarTWith, sampleRVarTWith
-    ) where
+    , runRVarT, sampleReaderRVarT, sampleStateRVarT
+    , runRVarTWith, sampleReaderRVarTWith, sampleStateRVarTWith
 
+    , RGen(..)
+    , uniformRVarT
+    , uniformRangeRVarT
+
+    , Prim(..)
+    ) where
 
-import Data.Random.Internal.Source (Prim(..), MonadRandom(..), RandomSource(..))
-import Data.Random.Source ({-instances-})
 
-import qualified Control.Monad.Trans.Class as T
-import Control.Monad (liftM, ap)
-import Control.Monad.Prompt (MonadPrompt(..), PromptT, runPromptT)
 import qualified Control.Monad.IO.Class as T
-import qualified Control.Monad.Trans as MTL
+import Control.Monad.Prompt (MonadPrompt(..), PromptT, runPromptT)
+import Control.Monad.Reader as MTL
+import Control.Monad.State as MTL
+import qualified Control.Monad.Trans.Class as T
 import qualified Data.Functor.Identity as T
+import Data.RVar.Prim
+import System.Random.Stateful
 
--- |An opaque type modeling a \"random variable\" - a value 
--- which depends on the outcome of some random event.  'RVar's 
+-- |An opaque type modeling a \"random variable\" - a value
+-- which depends on the outcome of some random event.  'RVar's
 -- can be conveniently defined by an imperative-looking style:
--- 
+--
 -- > normalPair =  do
 -- >     u <- stdUniform
 -- >     t <- stdUniform
 -- >     let r = sqrt (-2 * log u)
 -- >         theta = (2 * pi) * t
--- >         
+-- >
 -- >         x = r * cos theta
 -- >         y = r * sin theta
 -- >     return (x,y)
--- 
+--
 -- OR by a more applicative style:
--- 
+--
 -- > logNormal = exp <$> stdNormal
 --
 -- Once defined (in any style), there are several ways to sample 'RVar's:
--- 
+--
 -- * In a monad, using a 'RandomSource':
--- 
+--
 -- > runRVar (uniform 1 100) DevRandom :: IO Int
--- 
+--
 -- * In a monad, using a 'MonadRandom' instance:
 --
 -- > sampleRVar (uniform 1 100) :: State PureMT Int
--- 
+--
 -- * As a pure function transforming a functional RNG:
--- 
+--
 -- > sampleState (uniform 1 100) :: StdGen -> (Int, StdGen)
 --
 -- (where @sampleState = runState . sampleRVar@)
 type RVar = RVarT T.Identity
 
+-- | Sample random variable using `RandomGen` generator as source of entropy
+pureRVar :: RandomGen g => RVar a -> g -> (a, g)
+pureRVar rvar g = runStateGen g (runRVar rvar)
+
 -- |\"Run\" an 'RVar' - samples the random variable from the provided
 -- source of entropy.
-runRVar :: RandomSource m s => RVar a -> s -> m a
+runRVar :: StatefulGen g m => RVar a -> g -> m a
 runRVar = runRVarTWith (return . T.runIdentity)
 
 -- |@sampleRVar x@ is equivalent to @runRVar x 'StdRandom'@.
-sampleRVar :: MonadRandom m => RVar a -> m a
-sampleRVar = sampleRVarTWith (return . T.runIdentity)
+sampleReaderRVar :: (StatefulGen g m, MonadReader g m) => RVar a -> m a
```

Comment:
Think of this `mwc-random` instance from `random-source`: 
```haskell
(PrimMonad m, s ~ PrimState m) => MonadRandom (ReaderT (Gen s) m)
```

This is exactly what it would correspond to here with `StatefulGen`, i.e. one of resolved types for `sampleReaderRVar` could be:

```haskell
sampleReaderRVar :: RVar a -> ReaderT (Gen RealWorld) IO a
```

this would allow a user avoid threading the `Gen` itself, which is exactly what `MonadRandom` allowed

Reply:
Ah like this

```
MWC.create >>= runReaderT (sample (R.Normal 0.0 1.0) :: ReaderT (MWC.Gen RealWorld) IO Double)
```

But I then need the

```
instance (s ~ PrimState m, PrimMonad m) => StatefulGen (MWC.Gen s) m where
```

Reply:
That's right. Good thing it is now available :wink:

Reply:
https://hackage.haskell.org/package/mwc-random-0.15.0.1/docs/System-Random-MWC.html#t:Gen

Reply:
I just looked and I am using 0.14 :(